File: 2454-next-greater-element-iv/2454-next-greater-element-iv.java
class stack{
    int[]arr;
    int top=-1;

    stack(int n){
     this.arr=new int[n];
    }
    void push(int v){
        top+=1;
        this.arr[top]=v;
    }
    int pop(){
        return this.arr[this.top--];
    }
    boolean has(){
        return this.top>=0;
    }
    int peek(){
        return this.arr[this.top];
    }
    int size(){
        return this.top+1;
    }
}
class Solution {

    public int[] secondGreaterElement(int[] nums) {
        int n=nums.length;
        var s1=new stack(n);
        var s2=new stack(n);
        var ans=new int[n];
        Arrays.fill(ans,-1);

        for(int i=0;i<n;i++){
            while (s2.has() && nums[s2.peek()]<nums[i])
                 ans[s2.pop()]=nums[i];

            var temp=new stack(s1.size());

            while (s1.has() && nums[s1.peek()]<nums[i])
                 temp.push(s1.pop());
            
            while(temp.has())
                s2.push(temp.pop());
            s1.push(i);

        }
        return ans;

    }
}

File: 2454-next-greater-element-iv/2454-next-greater-element-iv.py
class Solution:
    def secondGreaterElement(self, nums: list[int]) -> list[int]:
        n=len(nums)
        ans=[-1]*n
        stack=[]
        stack1=[]

        for i,e in enumerate(nums):
            while stack1 and nums[stack1[-1]]<e:
                ans[stack1.pop()]=e
            temp=[]

            while stack and nums[stack[-1]]<e:
                temp.append(stack.pop())
            
            while temp:
                stack1.append(temp.pop())
            stack.append(i)
        return ans
        

        